src/vos/vos_dtx.c

```diff
@@ -935,14 +1306,155 @@ vos_dtx_check(daos_handle_t coh, struct dtx_id *dti)
 	return do_vos_dtx_check(coh, dti, NULL);
 }
 
-void
+int
 vos_dtx_commit_internal(struct vos_container *cont, struct dtx_id *dtis,
-			int count)
+			int count, daos_epoch_t epoch)
 {
-	int	i;
+	struct vos_cont_df		*cont_df = cont->vc_cont_df;
+	struct umem_instance		*umm = vos_cont2umm(cont);
+	struct vos_dtx_scm_blob		*dsb;
+	struct vos_dtx_scm_blob		*dsb_prev;
+	umem_off_t			 dsb_off;
+	struct vos_dtx_cmt_ent_df	*dce_df;
+	int				 slots = 0;
+	int				 cur = 0;
+	int				 rc = 0;
+	int				 rc1 = 0;
+	int				 i;
+	int				 j;
+
+	dsb = umem_off2ptr(umm, cont_df->cd_dtx_committed_tail);
+	if (dsb != NULL)
+		slots = dsb->dsb_cap - dsb->dsb_count;
+
+	if (slots == 0)
+		goto new_blob;
+
+	umem_tx_add_ptr(umm, &dsb->dsb_count, sizeof(dsb->dsb_count));
+
+again:
+	if (slots > count)
+		slots = count;
+
+	count -= slots;
+
+	if (slots > 1) {
+		D_ALLOC(dce_df, sizeof(*dce_df) * slots);
+		if (dce_df == NULL)
+			return -DER_NOMEM;
+	} else {
+		dce_df = &dsb->dsb_commmitted_data[dsb->dsb_count];
+	}
+
+	for (i = 0, j = 0; i < slots; i++, cur++) {
+		struct vos_dtx_cmt_ent	*dce = NULL;
+
+		rc = vos_dtx_commit_one(cont, &dtis[cur], epoch, &dce);
+		if (rc1 == 0)
+			rc1 = rc;
+
+		if (dce != NULL) {
+			if (slots == 1)
+				pmem_memcpy_nodrain(dce_df, &dce->dce_base,
+						    sizeof(*dce_df));
+			else
+				memcpy(&dce_df[j], &dce->dce_base,
+				       sizeof(dce_df[j]));
+			j++;
+		}
+	}
+
+	if (dce_df != &dsb->dsb_commmitted_data[dsb->dsb_count]) {
+		if (j > 0)
+			pmem_memcpy_nodrain(
+				&dsb->dsb_commmitted_data[dsb->dsb_count],
+				dce_df, sizeof(*dce_df) * j);
+		D_FREE(dce_df);
+	}
+
+	if (j > 0)
+		dsb->dsb_count += j;
+
+	if (count == 0)
+		return rc != 0 ? rc : rc1;
+
+	if (j < slots) {
+		slots -= j;
```

Comment:
In this case, is it possible to have "@cur >=  @count"?

Reply:
The @cur is the index for the dtx in the whole @dtis, the @count is the remaining dtx (to be committed) in the @dtis, so it is possible that "@Cur >= @count".